key.go

```diff
@@ -21,16 +22,30 @@ func BinaryFromDsKey(k datastore.Key) ([]byte, error) {
 	return base32.RawStdEncoding.DecodeString(k.String()[1:])
 }
 
+// MultihashToDsKey creates a Key from the given Multihash.
+func MultihashToDsKey(k mh.Multihash) datastore.Key {
+	return NewKeyFromBinary(k)
+}
+
+// DsKeyToMultihash converts a dsKey to the corresponding Multihash.
+func DsKeyToMultihash(dsKey datastore.Key) (mh.Multihash, error) {
+	kb, err := BinaryFromDsKey(dsKey)
+	if err != nil {
+		return nil, err
+	}
+	return mh.Cast(kb)
+}
+
 // CidToDsKey creates a Key from the given Cid.
 func CidToDsKey(k cid.Cid) datastore.Key {
-	return NewKeyFromBinary(k.Bytes())
+	return MultihashToDsKey(k.Hash())
 }
 
 // DsKeyToCid converts the given Key to its corresponding Cid.
 func DsKeyToCid(dsKey datastore.Key) (cid.Cid, error) {
-	kb, err := BinaryFromDsKey(dsKey)
+	hash, err := DsKeyToMultihash(dsKey)
 	if err != nil {
-		return cid.Cid{}, err
+		return cid.Cid{}, nil
 	}
-	return cid.Cast(kb)
+	return cid.NewCidV1(cid.Raw, hash), nil
```

Comment:
Do we want to change these? We should maintain them as they were and actually stop calling them in IPFS land. This is a breaking change for any application that relies on these functions (thinking Cluster), which will be unable to read data previously stored keyed by CID as `CidToDsKey` is producing a different Key now.

Reply:
What if we went with a go mod 2.0 release? I don't want to just leave these functions around doing the wrong thing.

Reply:
But they are doing what they say they are doing according to their name, which is using a CID as key (even if that's a bit nonsense).

If anything I'd prefer removing them directly. It's the best way to signal that a change happened and have users move to the new functions. Tagging a v2 means users will happily add `/v2` to their imports and things will "seem" to work as before because the functions are there and then break at runtime because of the underlying behaviour change.

I'd still keep them as they are, but otherwise I'd say remove + minor version tag (it's a utility function anyways).

Reply:
Personally, I'd expect these functions to map datastore keys to CIDs and back. We're just changing that mapping such that multiple CIDs now map to the same datastore key.

But I get what you're saying form a silent breakage standpoint. Given that, I'd go with your remove + minor tag version (to ensure we don't have two different versions at the same time). We'll have to update all callers but, honestly, that's probably for the best.

That'll also make the next migration slightly simpler (switching the blockstore interfaces over to CIDs).

Reply:
Given that you're handling the migration, want to take this?

Reply:
Yes